flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSelectiveSpillingStrategy.java

```diff
@@ -0,0 +1,115 @@
+    public static final float SPILL_BUFFER_RATIO = 0.4f;
+
+    /**
+     * When the number of buffers that have been requested exceeds this threshold, trigger the
+     * spilling operation.
+     */
+    public static final float SPILL_THRESHOLD = 0.7f;
+
+    /**
+     * For the case of buffer finished, there is no need to make a decision for {@link
+     * HsSelectiveSpillingStrategy}.
+     */
+    @Override
+    public Action onBufferFinished(
+            HsSpillingInfoProvider spillingInfoProvider, BufferWithIdentity finishedBuffer) {
+        return Action.EMPTY;
+    }
+
+    /**
+     * For the case of buffer consumed, this buffer need release. The control of the buffer is taken
+     * over by the downstream task.
+     */
+    @Override
+    public Action onBufferConsumed(
+            HsSpillingInfoProvider spillingInfoProvider, BufferWithIdentity consumedBuffer) {
+        return Action.onlyNeedRelease(Collections.singletonList(consumedBuffer));
+    }
+
+    /**
+     * When the amount of memory used exceeds the threshold, trigger the spilling operation and
+     * score the buffer of each subpartition. The lower the score, the more likely the buffer will
+     * be consumed in the next time, and should be kept in memory as much as possible. Select all
+     * buffers that need to be spilled according to the score from high to low.
+     */
+    @Override
+    public Action maybeSpill(HsSpillingInfoProvider spillInfoProvider) {
+        if (spillInfoProvider.getNumRequestedBuffers()
+                < spillInfoProvider.getPoolSize() * SPILL_THRESHOLD) {
+            return Action.EMPTY;
+        }
+
+        List<Integer> consumptionProgress = spillInfoProvider.getConsumptionProgress();
+        PriorityQueue<BufferIdentityWithScore> heap = new PriorityQueue<>();
+        int numBuffers = 0;
+        for (int channel = 0; channel < spillInfoProvider.getNumSubpartitions(); channel++) {
+            Collection<BufferWithIdentity> finishedBuffers =
+                    spillInfoProvider.getSubpartitionBuffers(channel);
+            int subpartitionProgress = consumptionProgress.get(channel);
+            // compute score for each subpartition buffers.
+            for (BufferWithIdentity bufferWithInfo : finishedBuffers) {
+                // buffer index in finishedBuffers must be greater than or equal to the downstream
+                // consumption progress, so the score must be greater than or equal to 0.
+                int score = bufferWithInfo.getBufferIndex() - subpartitionProgress;
+                heap.add(new BufferIdentityWithScore(bufferWithInfo, score));
+                numBuffers++;
+            }
+        }
+        List<BufferWithIdentity> toSpill = new ArrayList<>();
+        int spillNum = (int) (numBuffers * SPILL_BUFFER_RATIO);
+        for (int i = 0; i < spillNum; i++) {
+            toSpill.add(Preconditions.checkNotNull(heap.poll()).bufferWithInfo);
+        }
```

Comment:
Not sure about calculating the score for all the buffers. Consider we have a list/queue of buffers for each subpartition, it's probably good enough to always compare the first element of each queue (once the buffer is dequeued, reorder w.r.t. the new first element).

Reply:
It's really a nice idea, this can greatly improve the efficiency of the algorithm when there are a large number of buffers.